atef/bin/config.py

```diff
@@ -0,0 +1,1528 @@
+        """
+        try:
+            return cls._registry[data_type]
+        except KeyError:
+            ...
+        new_class = type(
+            f'QDataclassListFor{data_type.__name__}',
+            (cls, QObject),
+            {
+                'updated': QSignal(),
+                'added_value': QSignal(data_type),
+                'added_index': QSignal(int),
+                'removed_value': QSignal(data_type),
+                'removed_index': QSignal(int),
+                'changed_value': QSignal(data_type),
+                'changed_index': QSignal(int),
+            },
+        )
+        cls._registry[data_type] = new_class
+        return new_class
+
+    def get(self) -> List[Any]:
+        """
+        Return the current list of values.
+        """
+        return getattr(self.data, self.attr)
+
+    def append(self, new_value: Any) -> None:
+        """
+        Add a new value to the end of the list and update consumers.
+        """
+        data_list = self.get()
+        if data_list is None:
+            data_list = []
+            setattr(self.data, self.attr, data_list)
+        data_list.append(new_value)
+        self.added_value.emit(new_value)
+        self.added_index.emit(len(data_list) - 1)
+        self.updated.emit()
+
+    def remove_value(self, removal: Any) -> None:
+        """
+        Remove a value from the list by value and update consumers.
+        """
+        index = self.get().index(removal)
+        self.get().remove(removal)
+        self.removed_value.emit(removal)
+        self.removed_index.emit(index)
+        self.updated.emit()
+
+    def remove_index(self, index: int) -> None:
+        """
+        Remove a value from the list by index and update consumers.
+        """
+        value = self.get().pop(index)
+        self.removed_value.emit(value)
+        self.removed_index.emit(index)
+        self.updated.emit()
+
+    def put_to_index(self, index: int, new_value: Any) -> None:
+        """
+        Change a value in the list and update consumers.
+        """
+        self.get()[index] = new_value
+        self.changed_value.emit(new_value)
+        self.changed_index.emit(index)
+        self.updated.emit()
+
+
+class AtefCfgDisplay:
+    """Helper class for loading the .ui files and adding logic."""
+    filename: str
+
+    def __init_subclass__(cls):
+        """Read the file when the class is created"""
+        super().__init_subclass__()
+        cls.ui_form, _ = loadUiType(
+            str(Path(__file__).parent.parent / 'ui' / cls.filename)
+        )
+
+    def __init__(self, *args, **kwargs):
+        """Apply the file to this widget when the instance is created"""
+        super().__init__(*args, **kwargs)
+        self.ui_form.setupUi(self, self)
+
+    def retranslateUi(self, *args, **kwargs):
+        """Required function for setupUi to work in __init__"""
+        self.ui_form.retranslateUi(self, *args, **kwargs)
+
+
+class Window(AtefCfgDisplay, QMainWindow):
```

Comment:
QWidgets and screens -> atef.widgets (may need further submoduling)
Dataclass helpers -> atef.qt_helpers
stay in the script -> parser, main, and anything that's really script-specific?

Reply:
atef.widgets.base (the file loader + very general things)
atef.widgets.procedure
atef.widgets.config
maybe?

Reply:
Either way seems fine to me. I'm not sure if it's premature to break things up into a 'widgets' submodule or just forward-looking. I'd be OK leaving it to your discretion.

Reply:
I'm tempted to break it up simply because the widgets in these categories work together very tightly, and then not at all between categories- there is no relation between the atef config widgets and the procedure widgets, for example. So when I go to inspect the source code it will be easy to get confused looking in one massive "widgets" file.

Reply:
So this is a suggestion based on the present, not based on the future